osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerPlaylist.cs

```diff
@@ -0,0 +1,121 @@
+                {
+                    RelativeSizeAxes = Axes.X,
+                    Height = 25,
+                    Current = { BindTarget = DisplayMode }
+                },
+                new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Padding = new MarginPadding { Top = 27 },
+                    Masking = true,
+                    Children = new Drawable[]
+                    {
+                        queueList = new MultiplayerQueueList
+                        {
+                            RelativeSizeAxes = Axes.Both,
+                            SelectedItem = { BindTarget = SelectedItem }
+                        },
+                        historyList = new MultiplayerHistoryList
+                        {
+                            RelativeSizeAxes = Axes.Both,
+                            Alpha = 0,
+                            SelectedItem = { BindTarget = SelectedItem }
+                        }
+                    }
+                }
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            DisplayMode.BindValueChanged(onDisplayModeChanged, true);
+        }
+
+        private void onDisplayModeChanged(ValueChangedEvent<MultiplayerPlaylistDisplayMode> mode)
+        {
+            historyList.FadeTo(mode.NewValue == MultiplayerPlaylistDisplayMode.History ? 1 : 0, 100);
+            queueList.FadeTo(mode.NewValue == MultiplayerPlaylistDisplayMode.Queue ? 1 : 0, 100);
+        }
+
+        protected override void OnRoomUpdated()
+        {
+            base.OnRoomUpdated();
+
+            if (Room == null)
+            {
+                historyList.Items.Clear();
+                queueList.Items.Clear();
+                firstPopulation = true;
+                return;
+            }
+
+            if (firstPopulation)
+            {
+                foreach (var item in Room.Playlist)
+                    PlaylistItemAdded(item);
+
+                firstPopulation = false;
+            }
+        }
+
+        protected override void PlaylistItemAdded(MultiplayerPlaylistItem item)
+        {
+            base.PlaylistItemAdded(item);
+
+            var apiItem = Playlist.Single(i => i.ID == item.ID);
+
+            if (item.Expired)
+                historyList.Items.Add(apiItem);
+            else
+                queueList.Items.Add(apiItem);
+        }
+
+        protected override void PlaylistItemRemoved(long item)
+        {
+            base.PlaylistItemRemoved(item);
+
+            queueList.Items.RemoveAll(i => i.ID == item);
+            historyList.Items.RemoveAll(i => i.ID == item);
+        }
+
+        protected override void PlaylistItemChanged(MultiplayerPlaylistItem item)
+        {
+            base.PlaylistItemChanged(item);
+
+            PlaylistItemRemoved(item.ID);
+            PlaylistItemAdded(item);
```

Comment:
Seems super dodgy since this will be calling any base methods of those overrides. Are you okay if I split this out into a private method implementation?

Reply:
Sure, done.